src/Std/Data/HashSet/Basic.lean

```diff
@@ -0,0 +1,154 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Markus Himmel
+-/
+prelude
+import Std.Data.HashMap.Basic
+
+/-!
+# Hash sets
+
+This file develops the type `Std.Data.HashSet` of dependent hash sets.
+
+The operations `map` and `filterMap` on `Std.Data.HashSet` are defined in the module
+`Std.Data.HashSet.AdditionalOperations`.
+
+Lemmas about the operations on `Std.Data.HashSet` are available in the
+module `Std.Data.HashSet.Lemmas`.
+
+See the module `Std.Data.HashSet.Raw` for a variant of this type which is safe to use in
+nested inductive types.
+-/
+
+set_option linter.missingDocs true
+set_option autoImplicit false
+
+universe u v
+
+variable {α : Type u}
+
+namespace Std
+
+/-- Hash sets. -/
+structure HashSet (α : Type u) [BEq α] [Hashable α] where
+  /-- Internal implementation detail of the hash set. -/
+  inner : HashMap α Unit
+
+namespace HashSet
+
+/-- Creates a new empty hash set. The optional parameter `capacity` can be supplied to presize the
+map so that it can hold the given number of elements without reallocating. It is also possible to
+use the empty collection notations `∅` and `{}` to create an empty hash map with the default
+capacity. -/
+@[inline] def empty [BEq α] [Hashable α] (capacity := 8) : HashSet α :=
+  ⟨HashMap.empty capacity⟩
+
+instance [BEq α] [Hashable α] : EmptyCollection (HashSet α) where
+  emptyCollection := empty
+
+/-- Insert the given element into the set. -/
+@[inline] def insert [BEq α] [Hashable α] (m : HashSet α) (a : α) : HashSet α :=
+  ⟨m.inner.insertIfNew a ()⟩
+
+/-- Equivalent to (but potentially faster than) calling `contains` followed by `insert`. -/
+@[inline] def containsThenInsert [BEq α] [Hashable α] (m : HashSet α) (a : α) : Bool × HashSet α :=
+  let ⟨replaced, r⟩ := m.inner.containsThenInsertIfNew a ()
```

Comment:
But actually here it looks like it doesn't insert the new element. This strikes me as a footgun.

Reply:
I think this is the way it usually works with hash sets. [Rust](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.insert) and [Java](https://docs.oracle.com/en%2Fjava%2Fjavase%2F22%2Fdocs%2Fapi%2F%2F/java.base/java/util/HashSet.html#add(E)) do the same thing, except the boolean they return is flipped from how we do it.

Reply:
I think the docstring should mention this, then, just in case - I think there's perhaps a higher likelihood of someone trying to be clever here.

Reply:
I have updated the relevant docstrings to reflect this. Do you think the updated docstrings are clear enough?

Reply:
Yes! I think I would still be surprised, but then I would read the docstring, reflect, and think "ok, that makes sense" and feel good.